### ISSUE TITLE
Log-Freq Representations

### ISSUE THREAD

**Issue by stefan-balke** [librosa/librosa#394]
Hey all,

I'm not convinced by the output of all the CQT transforms for **analysis tasks**, maybe I'm too stupid. I'd like to add more log-freq representations to librosa.
Candidates would be:
- Straight binning approach (well, maybe we can leave this one out)
- Version using interpolation
- Reassigned Spectrogram
- Pitch-Filterbank form Müller et al. (using fixed coeffs from the chroma toolbox)

As I said, these are all for analysis tasks, since the representations are not invertible

My goal would be to have a resource where we could compare all these representations and immediately see, discuss the pros and cons...

Any thoughts on this?

**Comment by bmcfee** [comment]
Sure -- I'm all in favor of having more reference implementations of this kind of thing.

I'm not sure what "version using interpolation" means -- clarify?

Reassigned spectrogram shouldn't be too difficult, given that we already have `ifgram`, and plays into the remaining items of #12.

Pitch-filterbank -- say more about how this works?

**Comment by faroit** [comment]
> I'm not sure what "version using interpolation" means -- clarify?

@bmcfee, @stefan-balke, do you have something [like this](http://www.ee.columbia.edu/~dpwe/resources/matlab/sgram/logfmap.m) (@dpwe) in mind? I have a python version of this if that would be useful...

**Comment by stefan-balke** [comment]
`@bmcfee, @stefan-balke, do you have something like this (@dpwe) in mind? I have a python version of this if that would be useful...`

That's the binning strategy.

`Pitch-filterbank -- say more about how this works?`

That's what Meinard uses in the Chroma Toolbox. It's a filterbank similar to the CQT but only for analysis tasks. He's using elliptic filters and different downsampling factors, then doing some forward-backward convolution. He's describing it in his book _Information Retrieval for Music and Motion_.

`I'm not sure what "version using interpolation" means -- clarify?`

Very easy interpolation like this:

```
    f_stft = np.abs(librosa.stft(f_audio, center=True))

    f_stft_ax_freq_lin = librosa.fft_frequencies(sr=fs)
    f_stft_ax_freq_log = librosa.midi_to_hz(np.arange(1, 121))

    set_interp = interp1d(f_stft_ax_freq_lin, f_stft, kind='linear', axis=0)
    f_pitch = set_interp(f_stft_ax_freq_log)
```

**Comment by bzamecnik** [comment]
Hi, I just discovered librosa and I'm very happy to see a lot of techniques I previously coded myself in a nice audio processing package with comfortable API! Good job. Recently I published a python library [tfr](https://github.com/bzamecnik/tfr) to compute time-frequency reassigned spectrograms with requantization and mapping to pitch/chroma bins. In librosa I can see only `ifgram` (frequency reassignment) and all the spectrograms and ifgrams are binned in the input overlapping bins. 

Would anyone be interested in an effort to port the tfr to librosa? I could start hacking on that or anybody can join but first it would be good to discuss how to fit it nicely in the librosa the architecture.

Example outputs from tfr:

![comparison of reassignments](https://raw.githubusercontent.com/bzamecnik/tfr/master/examples/multicomponent_animation.gif)

- [Greensleeves - time-frequency reassigned pitchgram](https://www.youtube.com/watch?v=CQ1cjn_z_eQ)
- [Greensleeves intro - time-frequency reassigned chromagram in fifths ](https://www.youtube.com/watch?v=GX33y67CN-w)
- [Greensleeves - time-frequency reassigned chroma vector in polar plot](https://www.youtube.com/watch?v=DOJyjMQHP8U)

**Comment by bmcfee** [comment]
@bzamecnik Definitely! Thanks for the suggestion.

Refactoring `ifgram` and exposing an interface for time- and frequency-reassignment was also mentioned in #305 , but it's been a low-priority issue.

If you're interested in taking the lead on implementing this, it would be helpful if you could start a new issue describing what exactly the added functionality will be, and what parts of the API will need to change.  That way we can determine whether it should be scheduled for a major or minor revision.

**Comment by bzamecnik** [comment]
@bmcfee Great, thanks. First I need to get familiar with librosa (although I heard of it I've never used is so far) and think of what kind of changes have to be made. Of course it would call for a separate issue.

At the first sight it seems that the spectrograms and similar transforms are made on the overlapping input frames. In order to get good results with time-reassignment it's necessary to perform requantization and aggregation to non-overlapping output frames. Also frequency mapping (eg. to log-freq, pitch, mel scale) has to be done before the requantization. Sometimes we may be interested in the corrected positions without any quantization. So the overall process is a bit more complex. However the results are very nice.

**Comment by stefan-balke** [comment]
Hey @bzamecnik,

that looks great! Count me in for help, if you need some! I would start with a very general function, e.g. reassign a time-frequency representation to a arbitrary frequency and or time axis provided by the user. In the meantime I'll have some look at your code, which already looks promising.

**Comment by bzamecnik** [comment]
@stefan-balke Thanks! It would be certainly good to discuss the things, eg. the use cases etc. So far I tried to get the code into a usable state but I'd like to write an article explaining the method, how to works and how to use it in a simpler way than in the papers. I hope to get to this in the next few days.

**Comment by stefan-balke** [comment]
Partly solved by #439.
The only missing part for me would be reassignment. Already have some a notebook for that. Maybe later this year. But that can go in a separate issue imho.

**Comment by stefan-balke** [comment]
Anyone knowing works where temporal reassignment was actually used?

**Comment by bmcfee** [comment]
Reupping this: anyone feel like taking the lead on tf-reassignment?  At least scoping out what would need to be done so we can figure out how big of a project it is?

**Comment by bzamecnik** [comment]
Aah, yeah. I should port that code finally. It was a bit low priority in the last year and half...

**Comment by stefan-balke** [comment]
I'm still very interested and have some code already which works somehow but is not thorougly tested yet. Also, the reassignment is "jitted".

@bzamecnik @bmcfee any ideas for unit-tests? Putting in some test signals and see if the reassignment is within some limits?

As for API, I would imagine something like a `target_freqs` where simply the frequency values of the reassigned frequency axis are stored. Furthermore, it should take a complex valued STFT input or a time series...so very similar to the rest of the spectrogram-ish functions.

**Comment by stefan-balke** [comment]
Also, I'd like to have some kind of overview tutorial of the different TF-representations. Especially I'd like to learn about the Pros and Cons of the CQT... :)

**Comment by lostanlen** [comment]
@stefan-balke Can't go wrong with Sections 4.3 and 4.4 of the Wavelet Tour:
https://www.di.ens.fr/~mallat/papiers/WaveletTourChap1-2-3.pdf

Windowed Fourier Transform = STFT
Continuous Wavelet Transform = CQT
